File: frontend/lib/pricing.test.ts
import { describe, it, expect } from 'vitest';
import { estimateInvoicePrice, estimateDealerHoldback, calculateNetNetPrice, getTargetPriceType } from './pricing';

describe('estimateInvoicePrice', () => {
  it('should calculate invoice for Toyota sedan (lower margin)', () => {
    const result = estimateInvoicePrice(40000, 'Toyota', 'Sedan');
    expect(result).toBeCloseTo(37200, 0); // 40000 * (1 - 0.07) = 37200
  });

  it('should calculate invoice for Toyota truck (higher margin)', () => {
    const result = estimateInvoicePrice(40000, 'Toyota', 'Pickup Truck');
    expect(result).toBeCloseTo(36400, 0); // 40000 * (1 - 0.09) = 36400
  });

  it('should return MSRP for Tesla (direct-to-consumer)', () => {
    const result = estimateInvoicePrice(50000, 'Tesla', 'Sedan');
    expect(result).toBe(50000);
  });

  it('should use default margins for unknown brand', () => {
    const result = estimateInvoicePrice(40000, 'UnknownBrand', 'Sedan');
    expect(result).toBeCloseTo(38000, 0); // 40000 * (1 - 0.05) = 38000
  });

  it('should return 0 for zero MSRP', () => {
    const result = estimateInvoicePrice(0, 'Toyota', 'Sedan');
    expect(result).toBe(0);
  });

  it('should return 0 for negative MSRP', () => {
    const result = estimateInvoicePrice(-1000, 'Toyota', 'Sedan');
    expect(result).toBe(0);
  });

  it('should handle Honda SUV (higher margin)', () => {
    const result = estimateInvoicePrice(35000, 'Honda', 'SUV');
    expect(result).toBeCloseTo(31925, 0); // 35000 * (1 - 0.085) = 31925
  });
});

describe('estimateDealerHoldback', () => {
  it('should calculate holdback for Acura (2%)', () => {
    const result = estimateDealerHoldback(40000, 'Acura');
    expect(result).toBe(800); // 40000 * 0.02 = 800
  });

  it('should return 0 for Audi (no holdback)', () => {
    const result = estimateDealerHoldback(50000, 'Audi');
    expect(result).toBe(0);
  });

  it('should return 0 for BMW (no holdback)', () => {
    const result = estimateDealerHoldback(60000, 'BMW');
    expect(result).toBe(0);
  });

  it('should return 0 for Porsche (no holdback)', () => {
    const result = estimateDealerHoldback(100000, 'Porsche');
    expect(result).toBe(0);
  });

  it('should return 0 for Tesla (direct-to-consumer)', () => {
    const result = estimateDealerHoldback(50000, 'Tesla');
    expect(result).toBe(0);
  });

  it('should calculate holdback for Chevrolet (3%)', () => {
    const result = estimateDealerHoldback(40000, 'Chevrolet');
    expect(result).toBe(1200); // 40000 * 0.03 = 1200
  });

  it('should use default holdback for unknown brand', () => {
    const result = estimateDealerHoldback(40000, 'UnknownBrand');
    expect(result).toBe(800); // 40000 * 0.02 = 800
  });

  it('should return 0 for zero MSRP', () => {
    const result = estimateDealerHoldback(0, 'Toyota');
    expect(result).toBe(0);
  });
});

describe('calculateNetNetPrice', () => {
  it('should calculate net net for Toyota sedan', () => {
    const result = calculateNetNetPrice(40000, 'Toyota', 'Sedan');
    // Invoice: 37200, Holdback: 800, Net-Net: 36400
    expect(result).toBeCloseTo(36400, 0);
  });

  it('should return MSRP for Tesla (direct-to-consumer)', () => {
    const result = calculateNetNetPrice(50000, 'Tesla', 'Sedan');
    expect(result).toBe(50000);
  });

  it('should calculate net net for Chevrolet truck', () => {
    const result = calculateNetNetPrice(50000, 'Chevrolet', 'Pickup Truck');
    // Invoice: 46000 (8% max), Holdback: 1500, Net-Net: 44500
    expect(result).toBeCloseTo(44500, 0);
  });

  it('should return 0 for zero MSRP', () => {
    const result = calculateNetNetPrice(0, 'Toyota', 'Sedan');
    expect(result).toBe(0);
  });
});

describe('getTargetPriceType', () => {
  it('should return MSRP for Toyota Hybrid', () => {
    const result = getTargetPriceType('Toyota', 'Camry Hybrid');
    expect(result).toBe('MSRP');
  });

  it('should return MSRP for Porsche', () => {
    const result = getTargetPriceType('Porsche', '911');
    expect(result).toBe('MSRP');
  });

  it('should return Invoice for Honda', () => {
    const result = getTargetPriceType('Honda', 'Civic');
    expect(result).toBe('Invoice');
  });

  it('should return Invoice for Subaru', () => {
    const result = getTargetPriceType('Subaru', 'Outback');
    expect(result).toBe('Invoice');
  });

  it('should return Invoice for Mazda', () => {
    const result = getTargetPriceType('Mazda', 'CX-5');
    expect(result).toBe('Invoice');
  });

  it('should return Net-Net for Jeep', () => {
    const result = getTargetPriceType('Jeep', 'Wrangler');
    expect(result).toBe('Net-Net');
  });

  it('should return Net-Net for Ram', () => {
    const result = getTargetPriceType('Ram', '1500');
    expect(result).toBe('Net-Net');
  });

  it('should return Net-Net for Ford F-150', () => {
    const result = getTargetPriceType('Ford', 'F-150');
    expect(result).toBe('Net-Net');
  });

  it('should return Invoice as default for unknown combinations', () => {
    const result = getTargetPriceType('Unknown', 'Model');
    expect(result).toBe('Invoice');
  });
});

File: backend/internal/services/pricing_test.go
package services

import "testing"

func TestEstimateInvoicePrice(t *testing.T) {
	ps := NewPricingService()

	tests := []struct {
		name     string
		msrp     float64
		make     string
		bodyType string
		want     float64
	}{
		{
			name:     "Toyota sedan with 7-9% margin (should use lower end)",
			msrp:     40000,
			make:     "Toyota",
			bodyType: "Sedan",
			want:     37200, // 40000 * (1 - 0.07) = 37200
		},
		{
			name:     "Toyota truck with 7-9% margin (should use higher end)",
			msrp:     40000,
			make:     "Toyota",
			bodyType: "Pickup Truck",
			want:     36400, // 40000 * (1 - 0.09) = 36400
		},
		{
			name:     "Tesla direct-to-consumer (should return MSRP)",
			msrp:     50000,
			make:     "Tesla",
			bodyType: "Sedan",
			want:     50000,
		},
		{
			name:     "Unknown brand (should use default margins)",
			msrp:     40000,
			make:     "UnknownBrand",
			bodyType: "Sedan",
			want:     38000, // 40000 * (1 - 0.05) = 38000 (lower end of default)
		},
		{
			name:     "Zero MSRP",
			msrp:     0,
			make:     "Toyota",
			bodyType: "Sedan",
			want:     0,
		},
		{
			name:     "Honda with SUV body type",
			msrp:     35000,
			make:     "Honda",
			bodyType: "SUV",
			want:     32025, // 35000 * (1 - 0.085) = 32025 (higher end of 7-8.5%)
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := ps.EstimateInvoicePrice(tt.msrp, tt.make, tt.bodyType)
			if got != tt.want {
				t.Errorf("EstimateInvoicePrice(%v, %q, %q) = %v, want %v", tt.msrp, tt.make, tt.bodyType, got, tt.want)
			}
		})
	}
}

func TestEstimateDealerHoldback(t *testing.T) {
	ps := NewPricingService()

	tests := []struct {
		name string
		msrp float64
		make string
		want float64
	}{
		{
			name: "Acura with 2% holdback",
			msrp: 40000,
			make: "Acura",
			want: 800, // 40000 * 0.02 = 800
		},
		{
			name: "Audi with no holdback",
			msrp: 50000,
			make: "Audi",
			want: 0,
		},
		{
			name: "BMW with no holdback",
			msrp: 60000,
			make: "BMW",
			want: 0,
		},
		{
			name: "Porsche with no holdback",
			msrp: 100000,
			make: "Porsche",
			want: 0,
		},
		{
			name: "Tesla direct-to-consumer (no holdback)",
			msrp: 50000,
			make: "Tesla",
			want: 0,
		},
		{
			name: "Chevrolet with 3% holdback",
			msrp: 40000,
			make: "Chevrolet",
			want: 1200, // 40000 * 0.03 = 1200
		},
		{
			name: "Unknown brand (default 2% holdback)",
			msrp: 40000,
			make: "UnknownBrand",
			want: 800, // 40000 * 0.02 = 800
		},
		{
			name: "Zero MSRP",
			msrp: 0,
			make: "Toyota",
			want: 0,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := ps.EstimateDealerHoldback(tt.msrp, tt.make)
			if got != tt.want {
				t.Errorf("EstimateDealerHoldback(%v, %q) = %v, want %v", tt.msrp, tt.make, got, tt.want)
			}
		})
	}
}

func TestCalculateNetNetPrice(t *testing.T) {
	ps := NewPricingService()

	tests := []struct {
		name     string
		msrp     float64
		make     string
		bodyType string
		want     float64
	}{
		{
			name:     "Toyota sedan (invoice - holdback)",
			msrp:     40000,
			make:     "Toyota",
			bodyType: "Sedan",
			want:     36400, // Invoice: 37200, Holdback: 800, Net-Net: 36400
		},
		{
			name:     "Tesla direct-to-consumer (should return MSRP)",
			msrp:     50000,
			make:     "Tesla",
			bodyType: "Sedan",
			want:     50000,
		},
		{
			name:     "Chevrolet truck",
			msrp:     50000,
			make:     "Chevrolet",
			bodyType: "Pickup Truck",
			want:     44500, // Invoice: 46000 (using 8% max), Holdback: 1500, Net-Net: 44500
		},
		{
			name:     "Zero MSRP",
			msrp:     0,
			make:     "Toyota",
			bodyType: "Sedan",
			want:     0,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := ps.CalculateNetNetPrice(tt.msrp, tt.make, tt.bodyType)
			// Allow small floating point differences
			diff := got - tt.want
			if diff < 0 {
				diff = -diff
			}
			if diff > 1.0 {
				t.Errorf("CalculateNetNetPrice(%v, %q, %q) = %v, want %v (diff: %v)", tt.msrp, tt.make, tt.bodyType, got, tt.want, diff)
			}
		})
	}
}

